### ISSUE TITLE
Cocoon updated task with wrong status

### ISSUE THREAD

**Issue by CaseyHillers** [flutter/flutter#102998]
`Key(Checklist, 'flutter/engine/flutter-2.13-candidate.0/c5caf749fe75db788dba8def502c46c094435605', Task, 5151674742996992)` this task had a green LUCI build, but it was marked as failed in Cocoon.

I manually edited it in to have successful to unblock the release.

https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20iOS%20Engine%20Profile/10494/overview

Can someone investigate and see why this would have been marked red? I wonder if it was an issue on the LUCI side as we haven't had any recent changes.

**Comment by CaseyHillers** [comment]
This happened again when I merged a google roll CP in. The issue is https://ci.chromium.org/ui/p/flutter/builders/try/Mac%20iOS%20Engine/30647/overview isn't even related to the commit.

This may be related to https://github.com/flutter/cocoon/pull/1698 and incorrectly updating the task

**Comment by CaseyHillers** [comment]
My current investigation is that this would happen if the commit_key is null in https://github.com/flutter/cocoon/pull/1816/files#diff-44d26c271d034096ac3f0506b7400724ed6dd135aaff3b443ef628880a21457eL90

**Comment by CaseyHillers** [comment]
I put up https://github.com/flutter/cocoon/pull/1820 and will follow up later this week the tests necessary

**Comment by keyonghan** [comment]
Sounds related.
https://github.com/flutter/flutter/pull/103165 landed bunch of staging builders via cocoon scheduling, but it somehow confused prod builder statuses. For example: Linux build_tests_2_2 was updated with builder `#1, #4765`
<img width="579" alt="Screen Shot 2022-05-12 at 2 46 16 PM" src="https://user-images.githubusercontent.com/54558023/168174808-5f6ef99b-aa31-490c-9fec-e2cff97f5675.png">

/cc @yusuf-goog

**Comment by yusuf-goog** [comment]
@CaseyHillers believes having more builders just makes this problem more frequent.  Once he lands his fix, we should be able to reland https://github.com/flutter/flutter/pull/103165

**Comment by godofredoc** [comment]
@keyonghan @CaseyHillers which fix are we talking about here? did it land?

**Comment by godofredoc** [comment]
Seems like an async vs sync problem, is there a way to move the system to either fully asynchronous or synchronous?

**Comment by CaseyHillers** [comment]
The issue is https://github.com/flutter/flutter/issues/78876. Because of that, we need to write the entities to datastore to get the generated task key.

**Comment by keyonghan** [comment]
**tl;dr**: task and commit keys are valid, but cocoon scheduled `build` doesn't match corresponding `task and commit`.

**Observation**:
This happened only when a new target is added/enabled via cocoon schedule.
Example:
`Linux android views` was added in https://github.com/flutter/flutter/commit/9cfac48e2e8e54e28fec590ef9eed01639ab42f1 (without `scheduler: luci` means `cocoon schedule`)
This caused `cocoon scheduling` builds updated with wrong status:
`Linux build_tests_1_2` - builder list: `1292,4957`
`Linux firebase_abstract_method_smoke_test` - builder list `4730,6860`

Looking at builder `Linux build_tests_1_2`, the first scheduled build was with [`buildId= 8813269881498192433`](https://cr-buildbucket.appspot.com/rpcexplorer/services/buildbucket.v2.Builds/GetBuild?request={%20%20%20%20%22id%22:%20%228813269881498192353%22}), which actually was `build 1292` for builder `Linux android views`. However, the `task_key: 4931603798032384 ` maps to the real task `Linux build_tests_1_2`.

```
2022-05-24 08:58:20.151 PDT
Processing message 4681966095750136
2022-05-24 08:58:20.151 PDT
{"commit_key":"flutter/flutter/master/9cfac48e2e8e54e28fec590ef9eed01639ab42f1","task_key":"4931603798032384"}
2022-05-24 08:58:20.151 PDT
Updating buildId=8813269881498192433 for result=Result.failure
2022-05-24 08:58:20.151 PDT
Looking up key...
2022-05-24 08:58:20.198 PDT
Found Task(id: 4931603798032384, parentKey: flutter/flutter/master/9cfac48e2e8e54e28fec590ef9eed01639ab42f1, key: 4931603798032384, commitKey: flutter/flutter/master/9cfac48e2e8e54e28fec590ef9eed01639ab42f1, createTimestamp: 1653407888000, startTimestamp: 0, endTimestamp: 0, name: Linux build_tests_1_2, attempts: 1, isFlaky: false, isTestRunFlaky: false, timeoutInMinutes: 60, reason: , requiredCapabilities: [linux-vm], reservedForAgentId: , stageName: cocoon, status: In Progress, buildNumber: null, buildNumberList: null, builderName: Linux build_tests_1_2, luciBucket: null)
```

Seems something wrong with pubsub queue messaging when cocoon scheduling.

**Comment by keyonghan** [comment]
Another observation: when a cocoon-scheduled target is newly added, builds scheduled by cocoon were incorrectly using unrelated recipes. One example: https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20docs_test/4730/overview, which ran recipe `firebaselab/firebaselab`.

So, there are two issues:
1) Build is scheduled with incorrect builder config (recipe)
2) LUCI build is not correlated to the corresponding datastore task (build ID doesn't match task Key)

This affects the status only when a new target (scheduled by cocoon) is added. Lowering the priority to P3, and I am trying to repro.

**Comment by keyonghan** [comment]
Same thing happens on Engine Release branches: https://flutter-dashboard.appspot.com/#/build?taskFilter&authorFilter&messageFilter&hashFilter&showMac=true&showWindows=true&showiOS=true&showLinux=true&showAndroid=true&showStaging=false&repo=engine&branch=flutter-2.13-candidate.0

Tasks were mapped to non-related builds. One common thing with above mentioned issue is: there is a non-existent builder `Windows UWP Engine`. 

My guess is: this happens only when cocoon is scheduling a target that doesn't exist in LUCI config.

**Comment by keyonghan** [comment]
https://github.com/flutter/cocoon/pull/1917 is in.
@CaseyHillers Could you let me know when the next release change comes form engine repo? Interested to see how things go.

**Comment by CaseyHillers** [comment]
I'll add myself as an assignee to follow up with the results
